File: In/House/Q3/Solution.java
package In.House.Q3;

import java.util.*;

public class Solution {

    public int lengthOfLongestSubstring(String s) {
        int len = s.length();
        if (len <= 1) {
            return len;
        }
        Set<Character> set = new HashSet<>();
        int max = 0, start = 0;
        for (int i = 0; i < len; ++i) {
            char c = s.charAt(i);
            if (!set.contains(c)) {
                set.add(c);
                continue;
            }
            max = max < set.size() ? set.size(): max;
            while (true) {
                char cc = s.charAt(start ++);
                if (cc != c) {
                    set.remove(cc);
                    continue;
                }
                break;
            }
        }
        max = max < set.size() ? set.size(): max;
        return max;
    }

    public static void main(String[] args) {
        Solution slt = new Solution();
        System.out.println(slt.lengthOfLongestSubstring("abcabcbb")); // 3
        System.out.println(slt.lengthOfLongestSubstring("bbbbb")); // 1
        System.out.println(slt.lengthOfLongestSubstring("pwwkew")); // 3
        System.out.println(slt.lengthOfLongestSubstring("")); // 0
    }

}

File: In/House/Q3/solution.py
class Solution:
    # 1
    # def lengthOfLongestSubstring(self, s: str) -> int:
    #     if s == "":
    #         return 0
    #     rst = 1
    #     size = len(s)
    #     for idx in range(0, size - 1):
    #         if size - idx == rst:
    #             break
    #         st = {s[idx]}
    #         tmp = 1
    #         for ptr in range(idx + 1, size):
    #             if s[ptr] in st:
    #                 break
    #             tmp += 1
    #             st.add(s[ptr])
    #         rst = tmp if tmp > rst else rst
    #     return rst

    # 2
    def lengthOfLongestSubstring(self, s: str) -> int:
        if s == "":
            return 0
        rst = 1
        size = len(s)
        count = 1
        st = {}
        for idx in range(0, size - 1):
            if size - idx == rst:
                break
            if count == 1:
                st = {s[idx]}
                ptr = idx + 1
            else:
                st.remove(s[idx - 1])
                ptr = idx + count - 1
                count -= 1
            while ptr < size:
                if s[ptr] in st:
                    break
                count += 1
                st.add(s[ptr])
                ptr += 1
            rst = count if count > rst else rst
        return rst


slt = Solution()
print(slt.lengthOfLongestSubstring("abcabcbb"))  # 3
print(slt.lengthOfLongestSubstring("bbbbb"))  # 1
print(slt.lengthOfLongestSubstring("pwwkew"))  # 3
print(slt.lengthOfLongestSubstring("au"))  # 2
